### PR TITLE
Avoid repeatedly evaluating the source if it's a complex expression

### LINKED ISSUE #114: Slow compilation with multiple .modify calls when using Scala3
Original issue: https://github.com/lampepfl/dotty/issues/15945

Code to reproduce:

```scala
case class A(x: Double)
case class B(a1: A = A(1), a2: A = A(2), a3: A = A(3), a4: A = A(4))
case class C(b: B)

import com.softwaremill.quicklens._

@main
def main(): Unit = {
  val c = C(B(A(1)))
  c
    .modify(_.b.a1.x).setTo(0)
    .modify(_.b.a2.x).setTo(0)
    .modify(_.b.a3.x).setTo(0)
    .modify(_.b.a4.x).setTo(0)
}
```

cc @odersky @OndrejSpanel 

Maybe related to https://github.com/softwaremill/quicklens/pull/81 ?

### PR DESCRIPTION
**Pull request by adamw**
Closes #114

### CODE REVIEW

**Review by KacperFKorban — CHANGES_REQUESTED** [review]
Ooo that's a really interesting bug. Though I'm glad I wasn't the one reading the generated code snippets :sweat_smile:

I think that there's a more elegant solution for this. I'm pretty sure that simply removing `inline` here should fix the problem:
https://github.com/softwaremill/quicklens/blob/213a8e8eba0522e4fd229b6bafc07098a6c5d8f5/quicklens/src/main/scala-3/com/softwaremill/quicklens/package.scala#L10

**Comment by adamw** [comment]
@KacperFKorban Indeed! Nice :) I'm still stuck with Scala2-macro thinking ;) Seems to work - take another look then

**Review by OndrejSpanel — COMMENTED** [review]
Why are those tests in scala-3 specific folder? Would they do any harm in shared? Esp. the functionality tests seems application for Scala 2 as well.

**Review by KacperFKorban — APPROVED** [review]
LGTM

File: quicklens/src/test/scala-3/com/softwaremill/quicklens/test/CompileTimeTest.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package com.softwaremill.quicklens.test
+
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+class CompileTimeTest extends AnyFlatSpec with Matchers {
+  // #114
+  it should "not compile for too long in case of chained modify invocations" in {
+    val start = System.currentTimeMillis()
+    assertDoesNotCompile("""
```
Comment (OndrejSpanel) [inline quicklens/src/test/scala-3/com/softwaremill/quicklens/test/CompileTimeTest.scala:10]: While the code currently does not compile, would it really be a bug if it did? It does compile in Scala 2. I think `assertDoesNotCompile` should not be used here.

Comment (adamw) [reply quicklens/src/test/scala-3/com/softwaremill/quicklens/test/CompileTimeTest.scala:10]: it's in Scala3 sources as `assertDoesNotCompile` is a Scala3-only thing as far as I know (in ScalaTest).

As for this not compiling... I though that was the point showing the compiler timing issue (which is fixed here), but now I see there are two problems

Comment (adamw) [reply quicklens/src/test/scala-3/com/softwaremill/quicklens/test/CompileTimeTest.scala:10]: I've extracted this as a separate issue, https://github.com/softwaremill/quicklens/issues/116

Comment (adamw) [reply quicklens/src/test/scala-3/com/softwaremill/quicklens/test/CompileTimeTest.scala:10]: And changed the type here to an unrelated one, which definitely shouldn't work :)

**Pull request merged**